In [2]:
import os
import csv
from pathlib import Path

Prendo le classi disponibili dalla cartella da D1

In [13]:
# Percorso di D1
percorso_d1 = "Dataset/D1"  # Percorso relativo

def estrai_classi_cartelle(percorso):
    return [item.name for item in Path(percorso).iterdir() if item.is_dir()]

classi_d1 = estrai_classi_cartelle(percorso_d1)
print(len(classi_d1))
print(classi_d1)

200
['10247', '11090', '11211', '11212', '11214', '11458', '11476', '11477', '14704', '14719', '14769', '15068', '15070', '15100', '15379', '15392', '15535', '15573', '15712', '18651', '18654', '18674', '18677', '20482', '22388', '22885', '2357', '2412b', '2420', '24201', '24246', '2429', '2430', '2431', '2432', '2436', '2445', '2450', '2454', '2456', '24866', '25269', '2540', '26047', '2654', '26601', '26603', '26604', '2780', '27925', '28192', '2877', '3001', '3002', '3003', '3004', '3005', '3008', '3009', '3010', '30136', '3020', '3021', '3022', '3023', '3024', '3031', '3032', '3034', '3035', '3037', '30374', '3039', '3040', '30413', '30414', '3062b', '3065', '3068b', '3069b', '3070b', '32000', '32013', '32028', '32054', '32062', '32064', '32073', '32123', '32140', '32184', '32278', '32316', '3245c', '32523', '32524', '32525', '32526', '32607', '32952', '33291', '33909', '34103', '3460', '35480', '3622', '3623', '3660', '3665', '3666', '3673', '3700', '3701', '3705', '3710', '3713',

Prendo le classi da D2

Organizzo le classi in cartelle separate.

In [10]:
import shutil

# Percorso di D2
percorso = "Dataset/D2tofix/dataset"

def organizza_file_per_classe(percorso):
    percorso_base = Path(percorso)

    for file in percorso_base.iterdir():
        # Controlla se è un file
        if file.is_file():
            # Estrarre il codice della classe (prima dello spazio)
            codice_classe = file.name.split(' ')[0]
            # Percorso della cartella della classe
            cartella_classe = percorso_base / codice_classe

            # Crea la cartella se non esiste
            cartella_classe.mkdir(exist_ok=True)

            # Sposta il file nella cartella
            shutil.move(str(file), str(cartella_classe / file.name))
            print(f"Spostato '{file.name}' in '{cartella_classe}'")

organizza_file_per_classe(percorso)

Spostato '14719 flat tile corner 2x2 000L.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 000R.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 001L.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 001R.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 002L.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 002R.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 003L.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 003R.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 004L.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 004R.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 005L.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 flat tile corner 2x2 005R.png' in 'Dataset\D2tofix\dataset\14719'
Spostato '14719 

In [14]:
# Percorso di D2
percorso_d2 = "Dataset/D2tofix/dataset"

classi_d2 = estrai_classi_cartelle(percorso_d2)
print(len(classi_d2))
print(classi_d2)

50
['14719', '15672', '18654', '2357', '2420', '2780', '27925', '3001', '3002', '3003', '3004', '3005', '3010', '3020', '3021', '3022', '3023', '3024', '3037', '3038', '3039', '3040', '3045', '3046', '3062', '3063', '3068', '3069', '3070', '3298', '33909', '3622', '3623', '3659', '3675', '3700', '3794', '4150', '41677', '41678', '4274', '4286', '43093', '43857', '4490', '54200', '6143', '6632', '85984', '99301']


Controllo le classi in comune tra D1 e D2

In [16]:
classi_d1d2 = list(set(classi_d1) & set(classi_d2))
print(f"Classi in comune tra D1 e D2: {len(classi_d1d2)}")
print(classi_d1d2)

classi_d2md1 = list(set(classi_d2) - set(classi_d1))
print(f"Classi presenti in D2 ma non in D1: {len(classi_d2md1)} ")
print(classi_d2md1)

Classi in comune tra D1 e D2: 31
['3021', '3020', '3010', '3005', '4274', '2357', '3040', '3023', '3024', '3700', '54200', '2420', '18654', '3039', '3622', '3003', '85984', '3037', '3001', '2780', '33909', '14719', '43093', '3022', '6632', '4286', '27925', '3004', '3623', '3002', '41677']
Classi presenti in D2 ma non in D1: 19 
['3298', '6143', '43857', '3068', '4150', '3038', '3794', '3045', '3046', '41678', '3062', '3070', '3659', '3675', '3063', '3069', '15672', '4490', '99301']


Ci sono delle classi in D2 che non sono in D1.
Questo può essere un problema dato che D1 ha molte più istanze di D2.
Se decidessi di utilizzare D2 dovrei trovare il modo di bilanciare le classi in qualche modo.

Prendo ogni classe di D3 e rinomino le cartelle con le etichette relative.

In [3]:
mapping_file = Path("Dataset/piecesMapping.CSV").resolve()
percorso_cartelle = Path("Dataset/Cropped Images").resolve()

name_to_code = {}
with open(mapping_file, "r") as f:
    reader = csv.reader(f)
    next(reader)  # Salta l'intestazione
    for row in reader:
        name_to_code[row[0].lower()] = row[1]

# Rinominare le cartelle nel percorso specificato
def rinomina_cartelle(percorso, mapping):
    for nome_cartella in os.listdir(percorso):
        percorso_completo = os.path.join(percorso, nome_cartella)

        # Controlla che sia una cartella
        if os.path.isdir(percorso_completo):
            nome_cartella_lower = nome_cartella.lower()

            # Trova il codice corrispondente al nome
            if nome_cartella_lower in mapping:
                nuovo_nome = mapping[nome_cartella_lower]
                nuovo_percorso = os.path.join(percorso, nuovo_nome)

                # Rinomina la cartella
                os.rename(percorso_completo, nuovo_percorso)
                print(f"Cartella '{nome_cartella}' rinominata in '{nuovo_nome}'")
            else:
                print(f"Nessun codice trovato per '{nome_cartella}'")

# Esempio di utilizzo

rinomina_cartelle(percorso_cartelle, name_to_code)

Nessun codice trovato per '2357'
Nessun codice trovato per '2412b'
Nessun codice trovato per '2420'
Nessun codice trovato per '3001'
Nessun codice trovato per '3002'
Nessun codice trovato per '3003'
Nessun codice trovato per '3004'
Nessun codice trovato per '3005'
Nessun codice trovato per '3010'
Nessun codice trovato per '3020'
Nessun codice trovato per '3021'
Nessun codice trovato per '3022'
Nessun codice trovato per '3023'
Nessun codice trovato per '3024'
Nessun codice trovato per '3039'
Nessun codice trovato per '3622'
Nessun codice trovato per '3623'
Nessun codice trovato per '3710'
Nessun codice trovato per '54200'
Nessun codice trovato per '6141'


Prendo le classi di D3

In [17]:
# Percorso di D3
percorso_d3 = "Dataset/D3tofix/Cropped Images"

classi_d3 = estrai_classi_cartelle(percorso_d3)
print(len(classi_d3))
print(classi_d3)

20
['2357', '2412b', '2420', '3001', '3002', '3003', '3004', '3005', '3010', '3020', '3021', '3022', '3023', '3024', '3039', '3622', '3623', '3710', '54200', '6141']


Confronto le classi di D3 a quelle di D1

In [18]:
classi_d1d3 = list(set(classi_d1) & set(classi_d3))
print(f"Classi in comune tra D1 e D2: {len(classi_d1d3)}")
print(classi_d1d3)

classi_d3md1 = list(set(classi_d3) - set(classi_d1))
print(f"Classi presenti in D2 ma non in D1: {len(classi_d3md1)} ")
print(classi_d3md1)

Classi in comune tra D1 e D2: 19
['3021', '3020', '3010', '3005', '2357', '3023', '3024', '54200', '2420', '3039', '3622', '3003', '3001', '2412b', '3022', '3710', '3004', '3623', '3002']
Classi presenti in D2 ma non in D1: 1 
['6141']


Ho spostato le istanze di D2 e D3 nelle proprie cartelle Dataset/D2 e Dataset/D3

In [22]:
# percorsi definitivi
percorso_d1 = "Dataset/D1"
percorso_d2 = "Dataset/D2"
percorso_d3 = "Dataset/D3"

# classi definitive
classi_d1 = estrai_classi_cartelle(percorso_d1)
print(f"Classi D1: {len(classi_d1)}")
print(classi_d1)

classi_d2 = estrai_classi_cartelle(percorso_d2)
print(f"Classi D2: {len(classi_d2)}")
print(classi_d2)

classi_d3 = estrai_classi_cartelle(percorso_d3)
print(f"Classi D3: {len(classi_d3)}")
print(classi_d3)


Classi D1: 200
['10247', '11090', '11211', '11212', '11214', '11458', '11476', '11477', '14704', '14719', '14769', '15068', '15070', '15100', '15379', '15392', '15535', '15573', '15712', '18651', '18654', '18674', '18677', '20482', '22388', '22885', '2357', '2412b', '2420', '24201', '24246', '2429', '2430', '2431', '2432', '2436', '2445', '2450', '2454', '2456', '24866', '25269', '2540', '26047', '2654', '26601', '26603', '26604', '2780', '27925', '28192', '2877', '3001', '3002', '3003', '3004', '3005', '3008', '3009', '3010', '30136', '3020', '3021', '3022', '3023', '3024', '3031', '3032', '3034', '3035', '3037', '30374', '3039', '3040', '30413', '30414', '3062b', '3065', '3068b', '3069b', '3070b', '32000', '32013', '32028', '32054', '32062', '32064', '32073', '32123', '32140', '32184', '32278', '32316', '3245c', '32523', '32524', '32525', '32526', '32607', '32952', '33291', '33909', '34103', '3460', '35480', '3622', '3623', '3660', '3665', '3666', '3673', '3700', '3701', '3705', '371

Le classi in D3 sono tutte presenti in D1 tranne '6141'.
Potrei utilizzare questo dataset assieme a D1 però renderebbe più rappresentate le classi presenti.
Se decidessi di utilizzare D3 dovrei trovare un modo per bilanciare le classi.

Controllo a questo punto le classi in comune nei 3 dataset e quelle esclusive di ciascun dataset.

In [19]:
# Classi in comune tra tutti
classi_d1d2d3 = list(set(classi_d1) & set(classi_d2) & set(classi_d3))
print(f"Classi in comune tra D1, D2 e D3: {len(classi_d1d2d3)}")
print(classi_d1d2d3)
# Classi esclusive di d1
classi_d1excl = list(set(classi_d1) - set(classi_d2) - set(classi_d3))
print(f"Classi esclusive di D1: {len(classi_d1excl)}")
print(classi_d1excl)
# Classi esclusive di d2
classi_d2excl = list(set(classi_d2) - set(classi_d1) - set(classi_d3))
print(f"Classi esclusive di D1: {len(classi_d2excl)}")
print(classi_d2excl)
# Classi esclusive di d3
classi_d3excl = list(set(classi_d3) - set(classi_d1) - set(classi_d2))
print(f"Classi esclusive di D1: {len(classi_d3excl)}")
print(classi_d3excl)

Classi in comune tra D1, D2 e D3: 17
['3021', '3020', '54200', '3001', '3010', '2420', '3005', '3039', '3022', '3002', '3623', '3622', '3004', '2357', '3023', '3003', '3024']
Classi esclusive di D1: 167
['4070', '32054', '41740', '3035', '3705', '2450', '87994', '3065', '4081b', '4032', '43722', '43723', '24201', '6134', '48729', '2445', '50950', '87087', '32316', '99781', '87552', '59443', '99207', '11214', '32062', '44728', '32184', '87580', '18677', '63965', '2456', '85861', '87079', '26604', '3937', '11212', '32073', '3034', '32523', '11090', '2454', '40490', '4477', '32607', '15068', '4073', '61252', '42003', '32278', '92280', '2654', '15573', '3795', '11211', '6541', '87083', '60474', '15100', '60479', '60592', '3032', '10247', '99206', '2877', '49668', '15712', '32028', '22388', '3031', '47457', '88323', '26601', '11458', '15392', '26603', '3062b', '4085', '3245c', '6636', '15379', '60470', '18674', '28192', '4589', '41770', '60483', '22885', '6536', '34103', '15070', '3069b', '

Ci sono solo 17 classi in comune tra tutti i dataset. Ovviamente D1 è il più numeroso e il più popolato.
Sto pensando di utilizzare solo D1 per il training però mi dispiacerebbe non utilizzare gli altri dataset.

D1 ha una grande quantità di immagini, ma sono in formato 64x64.
Penso che la risoluzione possa rendere il training più leggero, ma potrebbe anche limitare la capacità di apprendere della CNN.
Gli altri dataset presentano comunque un'ottima qualità.